cuda/components/memory.cuh

```diff
@@ -15,761 +15,12 @@
-                     convert_generic_ptr_to_smem_ptr(ptr)),
-                 "d"(real_result), "d"(imag_result)
-                 : "memory");
-#else
-    asm volatile("st.relaxed.cta.shared.v2.f64 [%0], {%1, %2};" ::"r"(
-                     convert_generic_ptr_to_smem_ptr(ptr)),
-                 "d"(real_result), "d"(imag_result)
-                 : "memory");
-#endif
-}
-
-
-__device__ __forceinline__ thrust::complex<float> load_relaxed(
-    const thrust::complex<float>* ptr)
-{
-    float real_result;
-    float imag_result;
-#if __CUDA_ARCH__ < 700
-    asm volatile("ld.volatile.v2.f32 {%0, %1}, [%2];"
-                 : "=f"(real_result), "=f"(imag_result)
-                 : "l"(const_cast<thrust::complex<float>*>(ptr))
-                 : "memory");
-#else
-    asm volatile("ld.relaxed.gpu.v2.f32 {%0, %1}, [%2];"
-                 : "=f"(real_result), "=f"(imag_result)
-                 : "l"(const_cast<thrust::complex<float>*>(ptr))
-                 : "memory");
-#endif
-    return thrust::complex<float>{real_result, imag_result};
-}
-
-
-__device__ __forceinline__ void store_relaxed(thrust::complex<float>* ptr,
-                                              thrust::complex<float> result)
-{
-    auto real_result = result.real();
-    auto imag_result = result.imag();
-#if __CUDA_ARCH__ < 700
-    asm volatile("st.volatile.v2.f32 [%0], {%1, %2};" ::"l"(ptr),
-                 "f"(real_result), "f"(imag_result)
-                 : "memory");
-#else
-    asm volatile("st.relaxed.gpu.v2.f32 [%0], {%1, %2};" ::"l"(ptr),
-                 "f"(real_result), "f"(imag_result)
-                 : "memory");
-#endif
-}
-
-
-__device__ __forceinline__ thrust::complex<double> load_relaxed(
-    const thrust::complex<double>* ptr)
-{
-    double real_result;
-    double imag_result;
-#if __CUDA_ARCH__ < 700
-    asm volatile("ld.volatile.v2.f64 {%0, %1}, [%2];"
-                 : "=d"(real_result), "=d"(imag_result)
-                 : "l"(const_cast<thrust::complex<double>*>(ptr))
-                 : "memory");
-#else
-    asm volatile("ld.relaxed.gpu.v2.f64 {%0, %1}, [%2];"
-                 : "=d"(real_result), "=d"(imag_result)
-                 : "l"(const_cast<thrust::complex<double>*>(ptr))
-                 : "memory");
-#endif
-    return thrust::complex<double>{real_result, imag_result};
-}
-
-
-__device__ __forceinline__ void store_relaxed(thrust::complex<double>* ptr,
-                                              thrust::complex<double> result)
-{
-    auto real_result = result.real();
-    auto imag_result = result.imag();
-#if __CUDA_ARCH__ < 700
-    asm volatile("st.volatile.v2.f64 [%0], {%1, %2};" ::"l"(ptr),
-                 "d"(real_result), "d"(imag_result)
-                 : "memory");
-#else
-    asm volatile("st.relaxed.gpu.v2.f64 [%0], {%1, %2};" ::"l"(ptr),
-                 "d"(real_result), "d"(imag_result)
-                 : "memory");
-#endif
-}
+#include "common/cuda_hip/components/memory.nvidia.hpp.inc"
```

Comment:
I would also be fine with putting the header in `cuda/...`. Including that one from hip would be reasonable for me.

Reply:
it's only an incomplete header, the full header has namespaces around, so I believe this is cleaner, since it uses `common` for incomplete files

Reply:
I would also prefer staying in the common because it is used from both